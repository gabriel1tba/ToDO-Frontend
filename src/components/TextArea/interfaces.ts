export interface ITextArea
  extends React.TextareaHTMLAttributes<HTMLTextAreaElement> {
  name: string;
  error: string | undefined;
}
