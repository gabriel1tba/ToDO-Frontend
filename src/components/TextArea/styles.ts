import styled, { css } from 'styled-components';

import Tooltip from '../Tooltip';

export interface IWrapper {
  isFocused: boolean;
  isFilled: boolean;
  hasError: boolean;
}

export const Wrapper = styled.div<IWrapper>`
  display: flex;
  align-items: center;

  background: ${({ theme }) => theme.colors.background};
  border-radius: 10px;
  border: 1px solid ${({ theme }) => theme.colors.gray[500]};
  padding: 8px;
  width: 100%;
  color: ${({ theme }) => theme.colors.gray[500]};

  & + div {
    margin-top: 8px;
  }

  ${({ hasError }) =>
    hasError &&
    css`
      border-color: ${({ theme }) => theme.colors.danger.main};
    `}

  ${({ isFilled }) =>
    isFilled &&
    css`
      color: ${({ theme }) => theme.colors.primary.main};
      border-color: ${({ theme }) => theme.colors.primary.main};
    `}

    ${({ isFocused }) =>
    isFocused &&
    css`
      color: ${({ theme }) => theme.colors.primary.main};
      border-color: ${({ theme }) => theme.colors.primary.main};
    `}

  textarea {
    flex: 1;
    border: 0;
    background: transparent;
    font-size: 16px;

    resize: none;

    color: #202124;
    box-shadow: 0 0 0 30px ${({ theme }) => theme.colors.background} inset;

    &:-webkit-autofill {
      -webkit-text-fill-color: ${({ theme }) => theme.colors.primary.main};
    }

    &::placeholder {
      color: #202124;
    }
  }
`;

export const Error = styled(Tooltip)`
  height: 20px;
  margin-left: 16px;

  span {
    background: ${({ theme }) => theme.colors.danger.main};
    color: ${({ theme }) => theme.colors.background};
    font-size: 16px;

    &::before {
      border-color: ${({ theme }) => theme.colors.danger.main} transparent;
    }
  }
`;
