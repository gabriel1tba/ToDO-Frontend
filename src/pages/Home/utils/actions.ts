import { TTodo } from 'services/TodoService/interfaces';

export enum Types {
  GET_TODOS = 'GET_TODOS',
  CREATE_TODO = 'CREATE_TODO',
  UPDATE_TODO = 'UPDATE_TODO',
  DELETE_TODO = 'DELETE_TODO',
}

interface GetTodos {
  type: Types.GET_TODOS;
  payload: TTodo[];
}

interface CreateTodo {
  type: Types.CREATE_TODO;
  payload: TTodo;
}

interface UpdateTodo {
  type: Types.UPDATE_TODO;
  payload: TTodo;
}

interface DeleteTodo {
  type: Types.DELETE_TODO;
  payload: string;
}

export type Action = GetTodos | CreateTodo | UpdateTodo | DeleteTodo;

export const getTodosAction = (todos: TTodo[]): GetTodos => ({
  type: Types.GET_TODOS,
  payload: todos,
});

export const createTodoAction = (todo: TTodo): CreateTodo => ({
  type: Types.CREATE_TODO,
  payload: todo,
});

export const updateTodoAction = (todo: TTodo): UpdateTodo => ({
  type: Types.UPDATE_TODO,
  payload: todo,
});

export const deleteTodoAction = (id: string): DeleteTodo => ({
  type: Types.DELETE_TODO,
  payload: id,
});
