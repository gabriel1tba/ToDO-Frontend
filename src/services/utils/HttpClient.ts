import axios from 'axios';

const HttpClient = axios.create({
  baseURL: `${process.env.REACT_APP_API_HOST}:${process.env.REACT_APP_API_PORT}`,
});

export default HttpClient;
