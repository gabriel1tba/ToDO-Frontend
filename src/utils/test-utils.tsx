import { render, RenderOptions } from '@testing-library/react';
import { TodoProvider } from 'context/todos';

interface IWrapperProvider {
  children: React.ReactNode;
}

const WrapperProvider = ({ children }: IWrapperProvider) => (
  <TodoProvider>{children}</TodoProvider>
);

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(<WrapperProvider>{ui}</WrapperProvider>, { ...options });

export * from '@testing-library/react';
export { customRender as render };
