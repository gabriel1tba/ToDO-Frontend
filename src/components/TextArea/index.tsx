import { forwardRef, ForwardRefRenderFunction, useRef, useState } from 'react';
import { FiAlertCircle } from 'react-icons/fi';
import { useTheme } from 'styled-components';

import * as S from './styles';

import { ITextArea } from './interfaces';

const TextArea: ForwardRefRenderFunction<HTMLTextAreaElement, ITextArea> = (
  { name, error, defaultValue = '', ...rest },
  ref
) => {
  const theme = useTheme();

  const [isFocused, setIsFocued] = useState(false);
  const [isFilled, setIsFilled] = useState(false);

  const textAreaRef = useRef({} as HTMLTextAreaElement);

  const handleInputFocus = () => {
    setIsFocued(true);
  };

  const handleInputBlur = () => {
    setIsFocued(false);

    textAreaRef.current.value ? setIsFilled(true) : setIsFilled(false);
  };

  return (
    <S.Wrapper
      hasError={!!error}
      isFocused={isFocused}
      isFilled={isFilled}
      onFocus={handleInputFocus}
      onBlur={handleInputBlur}
    >
      <textarea defaultValue={defaultValue} ref={ref} name={name} {...rest} />
      {error && (
        <S.Error title={error}>
          <FiAlertCircle
            color={
              isFocused ? theme.colors.gray[500] : theme.colors.orange.main
            }
            size={20}
          />
        </S.Error>
      )}
    </S.Wrapper>
  );
};

export default forwardRef(TextArea);
