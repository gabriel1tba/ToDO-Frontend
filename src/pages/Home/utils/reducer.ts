import { TTodo } from 'services/TodoService/interfaces';

import { Types, Action } from './actions';

const todosReducer = (state: TTodo[], action: Action): TTodo[] => {
  switch (action.type) {
    case Types.GET_TODOS: {
      return action.payload;
    }

    case Types.CREATE_TODO: {
      return [...state, action.payload];
    }

    case Types.UPDATE_TODO: {
      return state.map((todo) => {
        if (action.payload.id === todo.id) {
          todo = { ...action.payload, created_at: todo.created_at };
        }
        return todo;
      });
    }

    case Types.DELETE_TODO: {
      return state.filter((todo) => {
        return todo.id !== action.payload;
      });
    }

    default:
      return state;
  }
};

export default todosReducer;
