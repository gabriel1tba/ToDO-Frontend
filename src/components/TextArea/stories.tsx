import { Story, Meta } from '@storybook/react';

import TextArea from '.';
import { ITextArea } from './interfaces';

export default {
  title: 'Form/TextArea',
  component: TextArea,
  args: {
    label: 'E-mail',
    name: 'email',
    initialValue: '',
    placeholder: 'john.cage@gmail.com',
  },
  argTypes: {
    name: {
      type: 'string',
    },
    error: {
      type: 'string',
    },
  },
} as Meta;

export const Default: Story<ITextArea> = (args) => <TextArea {...args} />;

export const withError: Story<ITextArea> = (args) => <TextArea {...args} />;

withError.args = {
  error: 'Ops...something is wrong',
};
