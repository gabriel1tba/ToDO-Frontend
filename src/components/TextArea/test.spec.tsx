import userEvent from '@testing-library/user-event';
import { render, screen } from 'utils/test-utils';

import TextArea from '.';

describe('<TextArea />', () => {
  it('should render the TextArea correctly with Label', () => {
    const { container } = render(
      <TextArea name="descrição" placeholder="Digite uma descrição" error="" />
    );

    expect(
      screen.getByPlaceholderText(/digite uma descrição/i)
    ).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should write in the TextArea', async () => {
    render(
      <TextArea name="descrição" error="" placeholder="Digite uma descrição" />
    );

    userEvent.type(
      screen.getByPlaceholderText(/digite uma descrição/i),
      'Testes finalizados'
    );

    expect(screen.getByPlaceholderText(/digite uma descrição/i)).toHaveValue(
      'Testes finalizados'
    );
  });

  it('should render erros', async () => {
    render(<TextArea name="descrição" error="A descrição é obrigatória" />);

    expect(screen.getByText(/a descrição é obrigatória/i)).toBeInTheDocument();
  });
});
