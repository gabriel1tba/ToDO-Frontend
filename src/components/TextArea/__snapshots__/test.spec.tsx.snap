// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextArea /> should render the TextArea correctly with Label 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFFFFF;
  border-radius: 10px;
  border: 1px solid #424242;
  padding: 8px;
  width: 100%;
  color: #424242;
}

.c0 + div {
  margin-top: 8px;
}

.c0 textarea {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border: 0;
  background: transparent;
  font-size: 1rem;
  resize: none;
  color: #202124;
  box-shadow: 0 0 0 30px #FFFFFF inset;
}

.c0 textarea:-webkit-autofill {
  -webkit-text-fill-color: #ff9000;
}

.c0 textarea::-webkit-input-placeholder {
  color: #202124;
}

.c0 textarea::-moz-placeholder {
  color: #202124;
}

.c0 textarea:-ms-input-placeholder {
  color: #202124;
}

.c0 textarea::placeholder {
  color: #202124;
}

<div
  class="c0"
>
  <textarea
    name="descrição"
    placeholder="Digite uma descrição"
  />
  
</div>
`;
